lrpar/src/lib/lex_api.rs

```diff
@@ -3,26 +3,39 @@
 use std::{cmp, error::Error, fmt, hash::Hash, marker};
 
 use cfgrammar::Span;
-use num_traits::{PrimInt, Unsigned};
+use num_traits::{AsPrimitive, PrimInt, Unsigned};
+
+pub trait LexerTypes: fmt::Debug
+where
+    usize: AsPrimitive<Self::StorageT>,
+{
+    type LexemeT: Lexeme<Self::StorageT>;
+    type StorageT: 'static + fmt::Debug + Hash + PrimInt + Unsigned;
+    type LexErrorT: LexError;
+}
 
 /// The base trait which all lexers which want to interact with `lrpar` must implement.
-pub trait Lexer<LexemeT: Lexeme<StorageT>, StorageT: Hash + PrimInt + Unsigned, LexErrorT: LexError>
+pub trait Lexer<LexerTypesT: LexerTypes>
+where
+    usize: AsPrimitive<LexerTypesT::StorageT>,
 {
     /// Iterate over all the lexemes in this lexer. Note that:
     ///   * The lexer may or may not stop after the first [LexError] is encountered.
     ///   * There are no guarantees about what happens if this function is called more than once.
     ///     For example, a streaming lexer may only produce [Lexeme]s on the first call.
-    fn iter<'a>(&'a self) -> Box<dyn Iterator<Item = Result<LexemeT, LexErrorT>> + 'a>;
+    fn iter<'a>(
+        &'a self,
+    ) -> Box<dyn Iterator<Item = Result<LexerTypesT::LexemeT, LexerTypesT::LexErrorT>> + 'a>;
 }
 
 /// A `NonStreamingLexer` is one that takes input in one go, and is then able to hand out
 /// substrings to that input and calculate line and column numbers from a [Span].
 pub trait NonStreamingLexer<
     'input,
-    LexemeT: Lexeme<StorageT>,
-    StorageT: Hash + PrimInt + Unsigned,
-    LexErrorT: LexError,
->: Lexer<LexemeT, StorageT, LexErrorT>
+    StorageT: 'static + Hash + PrimInt + Unsigned,
+    LexerTypesT: LexerTypes<StorageT = StorageT>,
+>: Lexer<LexerTypesT> where
+    usize: AsPrimitive<StorageT>,
```

Comment:
I believe this one is the last remaining instance of `<StorageT, LexerTypesT>`.
I had a look at doing the same for this as I had done for the others, but have run into some troubles with type inference there in the tests.

Reply:
I'll have a look.

Reply:
6b65f60 should do the trick (we no longer needed all those generics in the test suite, as we hardcode `u16` in various places, hence the bizarre warning saying "expected u16, got u16"!).

Reply:
Awesome, going to resolve this and overall happy with the patch now.

Reply:
Cool! I think it's just the inevitable missing bits in the docs that we need to think about now?

Reply:
Yeah, that sounds right to me.